tensornetwork/network.py

```diff
@@ -1213,6 +1215,48 @@ def check_correct(self, check_connected: bool = True) -> None:
     if check_connected:
       self.check_connected()
 
+  def save(self, path: str):
+    """Serialize the network to disk in hdf5 format.
+
+    Args:
+      path: path to folder where network is saved.
+    """
+    with h5py.File(path, 'w') as net_file:
+      net_file.create_dataset('backend', data=self.backend.name)
+      nodes_group = net_file.create_group('nodes')
+      edges_group = net_file.create_group('edges')
+
+      for node in self.nodes_set:
+        node_group = nodes_group.create_group(node.name)
+        self._save_node(node, node_group)
+
+        for edge in node.edges:
+          if edge.node1 == node:
+            edge_group = edges_group.create_group(edge.name)
+            self._save_edge(edge, edge_group)
+
+  @staticmethod
+  def _save_node(node: network_components.BaseNode, node_group: h5py.Group):
+    if hasattr(node, '_tensor'):
+      node_group.create_dataset('tensor', data=node._tensor)
```

Comment:
I agree, not too much of a fan either, and the only reason it is there is because of the CopyNode's lazy tensor.

Perhaps it can make sense to make the _save_node a method of the node classes? Then CopyNode and Node's methods would be slightly different. Main drawback here is that despite being a method of the Node's, it should not be accessed directly from them (but hey, it's private!). The save method would then additionally save the class name for each node (default to the regular node for instance) to use the correct load method later. The latter would also be node-type specific.

This feels overkill now, but if we start adding more nodes (symmetries), it might be cleaner in the long run. Also, I think most of it can be factored into a abstract method of BaseNode, which keeps the specific implementations neat. CopyNode would probably just take over the abstract method, Node would additionally add the tensor.

Not sure if we need to do the same for Edges?

WDYT? I can give it a try tomorrow (getting a bit late in Europe)

Reply:
OK, I'll give it a try. I'd be inclined to make also _load_node a class method of the nodes, any objections to that?

Reply:
Added first refactoring. Note I thought I needed the changes in #216 (in the end, I probably don't), so I started from that one. Should merge in #216 first.